src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunTypeSystemHelper.cs

```diff
@@ -0,0 +1,386 @@
+                        {
+                            // We can resolve to exact method
+                            method = constrainedType.GetMethodDescForInterfaceMethod((MetadataType)interfaceType, interfaceMethod, throwOnConflict: false);
+                            isExactMethodResolved = (method != null);
+                        }
+                    }
+
+                    if (!isExactMethodResolved)
+                    {
+                        // We couldn't resolve the interface statically
+                        // Notify the caller that it should use runtime lookup
+                        // Note that we can leave pMD incorrect, because we will use runtime lookup
+                        forceRuntimeLookup = true;
+                    }
+                }
+                else
+                {
+                    // If we can resolve the interface exactly then do so (e.g. when doing the exact 
+                    // lookup at runtime, or when not sharing generic code).
+                    if (constrainedCanonType.CanCastTo(interfaceType))
+                    {
+                        method = constrainedCanonType.GetMethodDescForInterfaceMethod((MetadataType)interfaceType, genInterfaceMethod, throwOnConflict: false);
+                        if (method == null)
+                        {
+                            Debug.WriteLine("TryResolveConstraintMethodApprox: failed to find method desc for interface method\n");
+                        }
+                    }
+                }
+            }
+            else if (genInterfaceMethod.IsVirtual)
+            {
+                method = constrainedType.FindVirtualFunctionTargetMethodOnObjectType(genInterfaceMethod);
+            }
+            else
+            {
+                // The method will be null if calling a non-virtual instance 
+                // methods on System.Object, i.e. when these are used as a constraint.
+                method = null;
+            }
+
+            if (method == null)
+            {
+                // Fall back to VSD
+                return null;
+            }
+
+            //#TryResolveConstraintMethodApprox_DoNotReturnParentMethod
+            // Only return a method if the value type itself declares the method, 
+            // otherwise we might get a method from Object or System.ValueType
+            if (!method.OwningType.IsValueType)
+            {
+                // Fall back to VSD
+                return null;
+            }
+
+            // We've resolved the method, ignoring its generic method arguments
+            // If the method is a generic method then go and get the instantiated descriptor
+            if (interfaceMethod.HasInstantiation)
+            {
+                method = method.MakeInstantiatedMethod(interfaceMethod.Instantiation);
+            }
+
+            Debug.Assert(method != null);
+            //assert(!pMD->IsUnboxingStub());
```

Comment:
That's a shame. I do like that in Crossgen, `IsUnboxingStub` and `IsInstantiatingStub` are available on a `MethodDesc` so we don't have to plumb it all over. Might be something to consider in the future.

Reply:
Agreed, that totally sucks.

Reply:
This was a conscious choice that David made based on experience with these `MethodDesc`s in the NUTC codebase. They tend do show up in unexpected places and cause headaches. This assert is a prime example of having to guard against them showing up, because bad things happen if they do.

Piping through flags makes parts of the codebase that don't need to deal with unboxing/instantiating stubs easier. If I was doing this, I would probably go for an enum or a wrapper struct, because two bools are easy to interchange and cause a bug.